core/src/subgraph/context.rs

```diff
@@ -114,6 +115,62 @@ impl<C: Blockchain, T: RuntimeHostBuilder<C>> IndexingContext<C, T> {
         .await
     }
 
+    pub async fn process_block(
+        &self,
+        logger: &Logger,
+        block_ptr: BlockPtr,
+        block_data: Box<[u8]>,
+        handler: String,
+        mut state: BlockState<C>,
+        proof_of_indexing: &SharedProofOfIndexing,
+        causality_region: &str,
+        debug_fork: &Option<Arc<dyn SubgraphFork>>,
+        subgraph_metrics: &Arc<SubgraphInstanceMetrics>,
+        instrument: bool,
+    ) -> Result<BlockState<C>, MappingError> {
+        let error_count = state.deterministic_errors.len();
+
+        if let Some(proof_of_indexing) = proof_of_indexing {
+            proof_of_indexing
+                .borrow_mut()
+                .start_handler(causality_region);
+        }
+
+        let start = Instant::now();
+        state = self
+            .instance
+            .hosts()
+            .first()
```

Comment:
Hosts exist map 1:1 to DataSources, on substreams we only allow one data source since it's triggered on every block so we basically only care about the first one. 

The code around the runner has all these points where context is required to understand what is going. It's something I would like to refactor at some point in the near future.

Reply:
Added a comment